distributed/shuffle/_core.py

```diff
@@ -275,3 +282,33 @@ def __str__(self) -> str:
 
     def __hash__(self) -> int:
         return hash(self.run_id)
+
+
+@dataclass
+class ShuffleSpec:
+    id: ShuffleId
+
+
+@dataclass
+class ShuffleRunSpec(Generic[_T_partition_id]):
+    id: ShuffleId
+    run_id: int
+    worker_for: dict[_T_partition_id, str]
+    output_workers: set[str]  # TODO: Is this necessary?
+
+
+@dataclass(eq=False)
+class SchedulerShuffleState:
+    _run_id_iterator: ClassVar[itertools.count] = itertools.count(1)
+
+    id: ShuffleId
+    run_id: int
```

Comment:
As mentioned below, perhaps:

```suggestion
    id: ShuffleId
    run_id: int = dataclasses.field(init=False, default_factory=partial(next, itertools.count(1))
```
And now subclasses are no longer responsible for ensuring that the sequence number is unique.